src/auto-evo/mutation_strategy/UpgradeOrganelle.cs

```diff
@@ -0,0 +1,42 @@
+﻿using System;
+using System.Collections.Frozen;
+using System.Collections.Generic;
+using System.Linq;
+using AutoEvo;
+
+internal class UpgradeOrganelle : IMutationStrategy<MicrobeSpecies>
+{
+    private readonly FrozenSet<OrganelleDefinition> allOrganelles;
+    private IComponentSpecificUpgrades upgrade;
+
+    public UpgradeOrganelle(Func<OrganelleDefinition, bool> criteria, IComponentSpecificUpgrades upgrade)
+    {
+        allOrganelles = SimulationParameters.Instance.GetAllOrganelles().Where(criteria).ToFrozenSet();
+        this.upgrade = upgrade;
+    }
+
+    public bool Repeatable => true;
+
+    public List<Tuple<MicrobeSpecies, float>>? MutationsOf(MicrobeSpecies baseSpecies, float mp, bool lawk,
+        Random random)
+    {
+        if (allOrganelles.Count == 0)
+        {
+            return null;
+        }
+
+        var mutated = new List<Tuple<MicrobeSpecies, float>>();
+
+        MicrobeSpecies newSpecies = (MicrobeSpecies)baseSpecies.Clone();
+
+        foreach (OrganelleTemplate organelle in newSpecies.Organelles.Where(x => allOrganelles.Contains(x.Definition)))
```

Comment:
This `.Where` should be written out manually instead of with LINQ as this is a hot path in auto-evo so the performance and memory use of LINQ is not preferable here.